Add kernel and userspace callstack contexts feature and tests

frdeso opened the pull request.

compudj commented on tests/utils/testapp/gen-syscall-events-callstack/gen-syscall-events-callstack.c:
```diff
@@ -0,0 +1,154 @@
+		goto error;
+	}
+
+	while(ready != 1) {
```
missing whitespace after while.

compudj commented on tests/utils/testapp/gen-syscall-events-callstack/gen-syscall-events-callstack.c:
```diff
@@ -0,0 +1,154 @@
+		goto error;
+	}
+
+	path=argv[1];
```
missing whitespaces before/after =

compudj commented on tests/utils/testapp/gen-syscall-events-callstack/gen-syscall-events-callstack.c:
```diff
@@ -0,0 +1,154 @@
+		return -1;
+	}
+
+	(void) close(ret);
```
always check close return value, if only just to print an error and abort.

compudj commented on tests/utils/testapp/gen-syscall-events-callstack/gen-syscall-events-callstack.c:
```diff
@@ -0,0 +1,154 @@
+int create_file(const char *path)
+{
+	static bool file_created = false;
+	int ret;
```
missing newline after variable declaration.

compudj commented on tests/utils/testapp/gen-syscall-events-callstack/gen-syscall-events-callstack.c:
```diff
@@ -0,0 +1,154 @@
+
+int create_file(const char *path)
+{
+	static bool file_created = false;
```
we keep static variables at the top of the program.

compudj commented on tests/regression/kernel/test_callstack:
```diff
@@ -0,0 +1,134 @@
+	while [ ! -f "$file_sync" ]; do
+		sleep 0.5
+	done
+
```
how do you know the program won't trigger the system call before tracing starts ?

frdeso replied:
Because the syscall is executed when the signal is received from the shell script which is done only after tracing has started.

compudj commented on tests/regression/kernel/test_callstack:
```diff
@@ -0,0 +1,134 @@
+	start_lttng_tracing_ok
+
+	# Send signal to trigger the syscall
+	kill -SIGUSR1 $PID
```
how do you know you don't kill the program before it issues a single system call ?

frdeso replied:
We wait for the program to create a synchronization file. When the program has done that. It means that it has registered its signal handler.  We then start the tracing and send the signal that triggers the the execution of the syscall.

compudj replied:
Ah, yes. Agreed that it works. I would be tempted to stay clear from signals here since it's straightforward to do. Can you use the same scheme as tests/utils/testapp/gen-ust-events ? See mainly create_file() and wait_on_file(). We should perhaps move those to a helper library for test applications. This way, both rendez-vous points (in each direction) end up being file-based.

frdeso replied:
I agree that it make more sense to use the same technique at both places.

frdeso replied:
Actually, we only need for the program to wait for the test script to be ready (pid tracker). There is no need for the program to notify the test script that it's ready since were not using signals anymore. So, only wait_on_file() is needed here.

compudj commented on tests/utils/testapp/gen-syscall-events-callstack/gen-syscall-events-callstack.c:
```diff
@@ -0,0 +1,154 @@
+		goto error;
+	}
+
+	/* Create a file to signify that the signal handler is registered */
```
you should look more into the existing test apps for synchronization. The points at which the rendez-vous points are placed are important, and there are race scenarios that are not covered here.

frdeso commented:
I updated the PR addressing the comments.

frdeso commented:
Actually, let me add a quick fix

frdeso commented:
Okay, done.

compudj commented on tests/utils/testapp/gen-ust-events/gen-ust-events.c:
```diff
@@ -126,7 +89,18 @@ int main(int argc, char **argv)
+			if (ret != 0) {
+				goto end;
+			}
+			else
```
please use } else {   coding style.

compudj commented on tests/utils/utils.c:
```diff
@@ -66,3 +74,55 @@ int usleep_safe(useconds_t usec)
+
+	ret = creat(path, S_IRWXU);
+	if (ret < 0) {
+		fprintf(stderr, "Failed to create file %s\n", path);
```
you could probably use perror instead ?

frdeso commented:
Fixed

frdeso commented:
I rebased and updated this PR. It reuses commit 44dd3b5 from PR #90. So depending which of the two branches is merged first, the other will need to be rebased.

jgalar commented:
Leaving a note for work to do before merging this:
- MI representation of this context 
- Save/load representation of this context
- There may be some documentation updates needed

frdeso commented:
Added MI support for the following commands:
```
lttng --mi=xml add-context -k -t callstack-kernel
lttng --mi=xml add-context -k -t callstack-user
lttng --mi=xml add-context --list
```
and the save and load commands. 
As for the doc, there doesn't seem to be documentation for the contexts themselves.

jgalar commented on tests/utils/testapp/gen-syscall-events-callstack/gen-syscall-events-callstack.c:
```diff
@@ -0,0 +1,106 @@
+        : "0"(__NR_gettid)
+        : "cc", "rcx", "r11", "memory"
+    );
+    return ret;
```
Can you add a x86 (32 bit) implementation of this syscall?
Also, we would need to disable this test if we're not building on x86/x64.

jgalar replied:
I merged your syscall-generating test app a while ago. I'm guessing we can re-use the approach taken there instead of using an assembly stub?

frdeso replied:
done.

jgalar commented on include/lttng/event.h:
```diff
@@ -142,6 +142,8 @@ enum lttng_event_context_type {
 	LTTNG_EVENT_CONTEXT_PREEMPTIBLE		= 17,
 	LTTNG_EVENT_CONTEXT_NEED_RESCHEDULE	= 18,
 	LTTNG_EVENT_CONTEXT_MIGRATABLE		= 19,
+	LTTNG_EVENT_CONTEXT_CALLSTACK_KERNEL	= 20,
```
Add a comment specifying on which architecture this is supported.

frdeso replied:
Done for LTTNG_EVENT_CONTEXT_CALLSTACK_USER

jgalar commented on src/bin/lttng/commands/add_context.c:
```diff
@@ -240,6 +242,8 @@ const struct ctx_opts {
 	{ "preemptible", CONTEXT_PREEMPTIBLE },
 	{ "need_reschedule", CONTEXT_NEED_RESCHEDULE },
 	{ "migratable", CONTEXT_MIGRATABLE },
+	{ "callstack-kernel", CONTEXT_CALLSTACK_KERNEL },
```
Since those contexts are only available on x86/x64 (to my knowledge?), we can add check for those archs
```
#if defined(__x86_64) || defined(__i386)
	{ "callstack-kernel", CONTEXT_CALLSTACK_KERNEL },
	{ "callstack-user", CONTEXT_CALLSTACK_USER },
#endif

This way we don't list an unsupported context on other architectures.

frdeso replied:
CONTEXT_CALLSTACK_KERNEL is available on most (all?) architectures.
Only the CONTEXT_CALLSTACK_USER is only available on x86(64bit and 32bit)
I made the change you suggested for CONTEXT_CALLSTACK_USER

jgalar commented on tests/regression/kernel/test_callstack:
```diff
@@ -0,0 +1,140 @@
+function lttng_track_pid()
+{
+	local PID=$1
+	lttng track -s "$SESSION_NAME" -k --pid="$PID" > /dev/null
```
See this diff to make sure this works when the client is not installed (the utils in `utils.sh` launch the client from the `builddir`).

```
diff --git a/tests/regression/kernel/test_callstack b/tests/regression/kernel/test_callstack
index 5ffa33818..056acce26 100755
--- a/tests/regression/kernel/test_callstack
+++ b/tests/regression/kernel/test_callstack
@@ -19,7 +19,7 @@ TEST_DESC="Kernel tracer - Callstack context"
 
 CURDIR=$(dirname "$0")/
 TESTDIR=$CURDIR/../..
-NUM_TESTS=18
+NUM_TESTS=22
 TEST_APP="$TESTDIR/utils/testapp/gen-syscall-events-callstack/gen-syscall-events-callstack"
 PARSE_CALLSTACK="$TESTDIR/utils/parse-callstack.py"
 
@@ -30,13 +30,13 @@ source "$TESTDIR/utils/utils.sh"
 
 function lttng_untrack_all()
 {
-	lttng untrack -s "$SESSION_NAME" --all --pid -k > /dev/null
+	lttng_untrack 0 "-s $SESSION_NAME --all --pid -k"
 }
 
 function lttng_track_pid()
 {
 	local PID=$1
-	lttng track -s "$SESSION_NAME" -k --pid="$PID" > /dev/null
+	lttng_track 0 "-s $SESSION_NAME -k --pid=$PID"
 }
 
 function run_workload()
```

frdeso replied:
done.

jgalar commented on tests/utils/parse-callstack.py:
```diff
@@ -0,0 +1,170 @@
+            recorded_callstack=extract_user_func_names(executable, raw_callstack)
+        elif cs_type in 'kernel':
+            recorded_callstack=extract_kernel_func_names(raw_callstack)
+        else:
```
The test fails on my machine. Looking at the resulting traces, I do see that a callstack is captured.
However, the addresses don't match any symbol in `/proc/kallsyms` directly. I don't know if this is because the behaviour of the stack capture code changed at some point.

I'm using `4.16.13-1-ARCH`.

The user callstack part fails too. It seems addr2line can't map the addresses back :confused:.

frdeso replied:
The problem was caused by the fact that on Archlinux (and Ubuntu 18.04) binaries are compiled with the -pie (position independent executable) option making the address resolution more complex. For the test, we turn off this optimization using -no-pie.

jgalar commented on tests/utils/parse-callstack.py:
```diff
@@ -0,0 +1,170 @@
+    recorded_callstack = set()
+
+    # Remove commas and split on spaces
+    ip_not_found = True
```
Unused variable.

frdeso replied:
fixed

jgalar commented on tests/utils/testapp/gen-syscall-events-callstack/gen-syscall-events-callstack.c:
```diff
@@ -40,13 +40,25 @@ long __attribute__ ((noinline))
+        : "cc", "edi", "esi", "memory"
+    );
+#else
+#error "Userspace callstack test not supported for this architecture."
```
This will fail the build on all non-x86/x64 platforms. We should remove that test app from the build on those platforms.
That can be achieved by removing the folder from the in the parent `Makefile.am`as done here https://github.com/lttng/lttng-tools/blob/master/tests/regression/Makefile.am#L34.

I'm not sure what's the best way to detect the target architecture from autoconf and export that... I'm looking into it.

frdeso replied:
You're totally right. Having a #error here is a bad idea. We should simply prevent this binary from being compiled on unsupported architectures.

compudj replied:
#error when arch is unsupported is fine. But it should come along with arch detection in makefiles. lttng-ust and liburcu both do it.

jgalar commented:
Merged, thanks!
